canonicalwebteam/discourse_docs/parsers.py

```diff
@@ -0,0 +1,288 @@
+
+            # Remove control emoji
+            notification_html = blockquote.encode_contents().decode("utf-8")
+            notification_html = re.sub(
+                r"^\n?<p([^>]*)>ⓘ +", r"<p\1>", notification_html
+            )
+
+            notification = notification_template.render(
+                notification_class="p-notification", contents=notification_html
+            )
+            blockquote.replace_with(
+                BeautifulSoup(notification, features="html.parser")
+            )
+
+    for warning in soup.findAll("img", title=":warning:"):
+        first_paragraph = warning.parent
+        blockquote = first_paragraph.parent
+        last_paragraph = blockquote.findChildren(recursive=False)[-1]
+
+        if first_paragraph.name == "p" and blockquote.name == "blockquote":
+            warning.decompose()
+
+            # Remove extra padding/margin
+            first_paragraph.attrs["class"] = "u-no-padding--top"
+            if last_paragraph.name == "p":
+                if "class" in last_paragraph.attrs:
+                    last_paragraph.attrs["class"] += " u-no-margin--bottom"
+                else:
+                    last_paragraph.attrs["class"] = "u-no-margin--bottom"
+
+            # Strip leading space
+            first_item = last_paragraph.contents[0]
+            first_item.replace_with(first_item.lstrip(' '))
+
+            notification = notification_template.render(
+                notification_class="p-notification--caution",
+                contents=blockquote.encode_contents().decode("utf-8"),
+            )
+
+            blockquote.replace_with(
+                BeautifulSoup(notification, features="html.parser")
+            )
+
+    return soup
+
+
+def get_preamble(soup, break_on_title):
+    """
+    Given a BeautifulSoup HTML document,
+    separate out the HTML at the start, up to
+    the heading defined in `break_on_title`,
+    and return it as a BeautifulSoup object
+    """
+
+    heading = soup.find(re.compile("^h[1-6]$"), text=break_on_title)
+
+    if not heading:
+        return soup
+
+    preamble_elements = heading.fetchPreviousSiblings()
+    preamble_elements.reverse()
+    preamble_html = "".join(map(str, preamble_elements))
+
+    return BeautifulSoup(preamble_html, features="html.parser")
+
+
+def get_section(soup, title_text):
+    """
+    Given some HTML soup and the text of a title within it,
+    get the content between that title and the next title
+    of the same level, and return it as another soup object.
+    
+    E.g. if `soup` contains is:
+
+      <p>Pre</p>
+      <h2>My heading</h2>
+      <p>Content</p>
+      <h2>Next heading</h2>
+
+    and `title_text` is "My heading", then it will return:
+
+    <p>Content</p>
+    """
+
+    heading = soup.find(re.compile("^h[1-6]$"), text=title_text)
+
+    if not heading:
+        return None
+
+    heading_tag = heading.name
+
+    section_html = "".join(map(str, heading.fetchNextSiblings()))
```

Comment:
Do we need to go into `str` and then parse to HTML again?

What if `fetchNextSiblings` returns and empty list or none of its elements contains the `heading_tag`)? Could we not break directly and only trigger the parsing etc. if we need to?

Reply:
> Do we need to go into str and then parse to HTML again?

I don't think there's a neater way to do this than converting to string and back again. If you can find one then that'd be awesome. A good starting point is here: https://stackoverflow.com/questions/36076052/beautifulsoup-find-all-on-bs4-element-resultset-object-or-list.

> What if fetchNextSiblings returns and empty list

In this case, it will produce an empty string, which I think is valid:

``` python
> "".join(map(str, BeautifulSoup("<h1>hello</h1>", features="html.parser").fetchNextSiblings()))
''
```

> or none of its elements contains the heading_tag)?

I believe that's covered effectively by line 281 below.

``` python
> BeautifulSoup('', features="html.parser").find('h1') == None
True
```

So if the section is empty, it will return an empty string, which seems accurate.

Reply:
I refactored the code here ever-so-slightly to avoid parsing HTML into soup twice in some circumstances.

Reply:
I thought this might work:
```python
    heading_tag = heading.name

    heading_siblings = heading.fetchNextSiblings()


    # If there's another heading of the same level
    # get the content before it
    next_heading = heading_siblings.find(heading_tag)

    if next_heading:
        section_elements = next_heading.fetchPreviousSiblings()
        return section_elements.reverse()

    return heading_siblings
```

Reply:
That doesn't work directly, because this function is supposed to return a `BeautifulSoup` object, rather than a `ResultSet` - `heading_siblings` would be a `ResultSet`.

I did already restructured to avoid `fetchNextSiblings` when I might not need it. But I don't think this can get any neater unfortunately.